cpp/src/replace/replace.cu

```diff
@@ -13,26 +16,551 @@
+  int gridsz = gridDim.x;
+
+  int start = tid + blkid * blksz;
+  int step = blksz * gridsz;
+
+  for (int i=start; i<size; i+=step) {
+    out_data[i] = cudf::bit_is_set(in_valid, i)? in_data[i] : *replacement;
+  }
+}
+
+
+template <typename Type>
+__global__
+void replace_nulls_with_column(cudf::size_type size,
+                               Type const* __restrict__ in_data,
+                               cudf::bitmask_type const* __restrict__ in_valid,
+                               Type const* __restrict__ replacement,
+                               Type* __restrict__ out_data)
+{
+  int tid = threadIdx.x;
+  int blkid = blockIdx.x;
+  int blksz = blockDim.x;
+  int gridsz = gridDim.x;
+
+  int start = tid + blkid * blksz;
+  int step = blksz * gridsz;
+
+  for (int i=start; i<size; i+=step) {
+    out_data[i] = cudf::bit_is_set(in_valid, i)? in_data[i] : replacement[i];
+  }
+}
+
+
+/* --------------------------------------------------------------------------*/
+/**
+ * @brief Functor called by the `type_dispatcher` in order to invoke and instantiate
+ *        `replace_nulls` with the appropriate data types.
+ */
+/* ----------------------------------------------------------------------------*/
+struct replace_nulls_column_kernel_forwarder {
+  template <typename col_type>
+  void operator()(cudf::column_view const& input,
+                  cudf::column_view const& replacement,
+                  cudf::mutable_column_view& output,
+                  cudaStream_t stream = 0)
+  {
+    cudf::size_type nrows = input.size();
+    cudf::util::cuda::grid_config_1d grid{nrows, BLOCK_SIZE};
+
+    replace_nulls_with_column<<<grid.num_blocks, BLOCK_SIZE, 0, stream>>>(nrows,
+                                                                          input.data<col_type>(),
+                                                                          input.null_mask(),
+                                                                          replacement.data<col_type>(),
+                                                                          output.data<col_type>());
+
+  }
+};
+
+template<>
+void replace_nulls_column_kernel_forwarder::operator ()<cudf::string_view>(cudf::column_view const& input,
+                                                                           cudf::column_view const& replacement,
+                                                                           cudf::mutable_column_view& output,
+                                                                           cudaStream_t stream){
+  CUDF_FAIL("Strings not supported for replacement.");
+}
+
+
+/* --------------------------------------------------------------------------*/
+/**
+ * @brief Functor called by the `type_dispatcher` in order to invoke and instantiate
+ *        `replace_nulls` with the appropriate data types.
+ */
+/* ----------------------------------------------------------------------------*/
+struct replace_nulls_scalar_kernel_forwarder {
```

Comment:
There is a significant amount of redundancy between the `replace_nulls_column` and `replace_nulls_scalar`. This can all be vastly simplified by instead accepting an iterator for the replacement. For the `scalar` case, you can just use a `thrust::constant_iterator`, and for the column case you can just use `column_view::begin<T>()`.

Reply:
Done.